intellij-plugin-verifier/verifier-core/src/main/java/com/jetbrains/pluginverifier/verifiers/method/KotlinMethods.kt

```diff
@@ -122,6 +132,14 @@ object KotlinMethods {
 
     val actualKotlinOwner = methodInsnNode.owner.substringBeforeLast("\$DefaultImpls")
 
+    // If the current class is the default implementation of a child interface
+    // then kotlin make it the inner class of 2 inner classes
```

Comment:
Maybe I don't understand what you're saying, but a class cannot be inner in 2 classes. If you mean the number of `InnerClasses` attribute entries, then you should filter it anyway, because the `InnerClasses` attribute can list (and does so, [starting from Kotlin 1.8](https://youtrack.jetbrains.com/issue/KT-27936/Write-InnerClasses-attribute-for-all-class-names-used-in-a-class-file)) any classes which are inner/nested and which are _referenced_ in any way in the class file. So, for example, if the interface method happens to take `java.util.Map.Entry` as a parameter, there will be an `InnerClasses` entry for that.

Reply:
That's indeed what I'm referring to, i.e. the `InnerClasses` attributes.

So I was maybe extra cautious by adding this check, but the previous check is enough to distinguish these methods.

-----

I was a bit puzzled as I wasn't aware this class attribute could be used for any class references (additionally to expected inner / nested refs). The generated bytecode looks like this :

```
$ javap -s -c -p -l -v -constants -cp intellij-plugin-verifier/verifier-test/mock-plugin/build/classes/kotlin/main/ mock.plugin.internal.defaultMethod.NoInternalTypeUsageInterface\$DefaultImpls
Classfile /Users/brice.dutheil/opensource/intellij-plugin-verifier/intellij-plugin-verifier/verifier-test/mock-plugin/build/classes/kotlin/main/mock/plugin/internal/defaultMethod/NoInternalTypeUsageInterface$DefaultImpls.class
  Last modified Jul 24, 2023; size 2447 bytes
  MD5 checksum 2c1fec6a0c6d90e07c1b30a16950ab53
  Compiled from "NoInternalTypeUsageInterface.kt"
public final class mock.plugin.internal.defaultMethod.NoInternalTypeUsageInterface$DefaultImpls
  minor version: 0
  major version: 55
  flags: (0x0031) ACC_PUBLIC, ACC_FINAL, ACC_SUPER
  this_class: #2                          // mock/plugin/internal/defaultMethod/NoInternalTypeUsageInterface$DefaultImpls
  super_class: #4                         // java/lang/Object
  interfaces: 0, fields: 0, methods: 4, attributes: 3

  // skipped for brevity

}
InnerClasses:
  public static final #51= #22 of #20;    // DefaultImpls=class internal/defaultMethod/InterfaceWithDefaultMethodUsingInternalAPI$DefaultImpls of class internal/defaultMethod/InterfaceWithDefaultMethodUsingInternalAPI
  public static final #51= #2 of #53;     // DefaultImpls=class mock/plugin/internal/defaultMethod/NoInternalTypeUsageInterface$DefaultImpls of class mock/plugin/internal/defaultMethod/NoInternalTypeUsageInterface
SourceFile: "NoInternalTypeUsageInterface.kt"
RuntimeVisibleAnnotations:
  0: #43(#44=[I#45,I#46,I#45],#47=I#48,#49=I#50)
    kotlin.Metadata(
      mv=[1,5,1]
      k=3
      xi=48
    )
```




<details><summary>Complete `javap` output</summary>

```
Classfile /Users/brice.dutheil/opensource/intellij-plugin-verifier/intellij-plugin-verifier/verifier-test/mock-plugin/build/classes/kotlin/main/mock/plugin/internal/defaultMethod/NoInternalTypeUsageInterface$DefaultImpls.class
  Last modified Jul 24, 2023; size 2447 bytes
  MD5 checksum 2c1fec6a0c6d90e07c1b30a16950ab53
  Compiled from "NoInternalTypeUsageInterface.kt"
public final class mock.plugin.internal.defaultMethod.NoInternalTypeUsageInterface$DefaultImpls
  minor version: 0
  major version: 55
  flags: (0x0031) ACC_PUBLIC, ACC_FINAL, ACC_SUPER
  this_class: #2                          // mock/plugin/internal/defaultMethod/NoInternalTypeUsageInterface$DefaultImpls
  super_class: #4                         // java/lang/Object
  interfaces: 0, fields: 0, methods: 4, attributes: 3
Constant pool:
   #1 = Utf8               mock/plugin/internal/defaultMethod/NoInternalTypeUsageInterface$DefaultImpls
   #2 = Class              #1             // mock/plugin/internal/defaultMethod/NoInternalTypeUsageInterface$DefaultImpls
   #3 = Utf8               java/lang/Object
   #4 = Class              #3             // java/lang/Object
   #5 = Utf8               internalArgsReturningInternal
   #6 = Utf8               (Lmock/plugin/internal/defaultMethod/NoInternalTypeUsageInterface;Linternal/defaultMethod/AnInternalType;Ljava/lang/String;)Linternal/defaultMethod/AnInternalType;
   #7 = Utf8               Lorg/jetbrains/annotations/Nullable;
   #8 = Utf8               Lorg/jetbrains/annotations/NotNull;
   #9 = Utf8               anInternalType
  #10 = String             #9             // anInternalType
  #11 = Utf8               kotlin/jvm/internal/Intrinsics
  #12 = Class              #11            // kotlin/jvm/internal/Intrinsics
  #13 = Utf8               checkNotNullParameter
  #14 = Utf8               (Ljava/lang/Object;Ljava/lang/String;)V
  #15 = NameAndType        #13:#14        // checkNotNullParameter:(Ljava/lang/Object;Ljava/lang/String;)V
  #16 = Methodref          #12.#15        // kotlin/jvm/internal/Intrinsics.checkNotNullParameter:(Ljava/lang/Object;Ljava/lang/String;)V
  #17 = Utf8               s
  #18 = String             #17            // s
  #19 = Utf8               internal/defaultMethod/InterfaceWithDefaultMethodUsingInternalAPI
  #20 = Class              #19            // internal/defaultMethod/InterfaceWithDefaultMethodUsingInternalAPI
  #21 = Utf8               internal/defaultMethod/InterfaceWithDefaultMethodUsingInternalAPI$DefaultImpls
  #22 = Class              #21            // internal/defaultMethod/InterfaceWithDefaultMethodUsingInternalAPI$DefaultImpls
  #23 = Utf8               (Linternal/defaultMethod/InterfaceWithDefaultMethodUsingInternalAPI;Linternal/defaultMethod/AnInternalType;Ljava/lang/String;)Linternal/defaultMethod/AnInternalType;
  #24 = NameAndType        #5:#23         // internalArgsReturningInternal:(Linternal/defaultMethod/InterfaceWithDefaultMethodUsingInternalAPI;Linternal/defaultMethod/AnInternalType;Ljava/lang/String;)Linternal/defaultMethod/AnInternalType;
  #25 = Methodref          #22.#24        // internal/defaultMethod/InterfaceWithDefaultMethodUsingInternalAPI$DefaultImpls.internalArgsReturningInternal:(Linternal/defaultMethod/InterfaceWithDefaultMethodUsingInternalAPI;Linternal/defaultMethod/AnInternalType;Ljava/lang/String;)Linternal/defaultMethod/AnInternalType;
  #26 = Utf8               $this
  #27 = Utf8               Lmock/plugin/internal/defaultMethod/NoInternalTypeUsageInterface;
  #28 = Utf8               Linternal/defaultMethod/AnInternalType;
  #29 = Utf8               Ljava/lang/String;
  #30 = Utf8               internalArgsReturningVoid
  #31 = Utf8               (Lmock/plugin/internal/defaultMethod/NoInternalTypeUsageInterface;Linternal/defaultMethod/AnInternalType;Ljava/lang/String;)V
  #32 = Utf8               (Linternal/defaultMethod/InterfaceWithDefaultMethodUsingInternalAPI;Linternal/defaultMethod/AnInternalType;Ljava/lang/String;)V
  #33 = NameAndType        #30:#32        // internalArgsReturningVoid:(Linternal/defaultMethod/InterfaceWithDefaultMethodUsingInternalAPI;Linternal/defaultMethod/AnInternalType;Ljava/lang/String;)V
  #34 = Methodref          #22.#33        // internal/defaultMethod/InterfaceWithDefaultMethodUsingInternalAPI$DefaultImpls.internalArgsReturningVoid:(Linternal/defaultMethod/InterfaceWithDefaultMethodUsingInternalAPI;Linternal/defaultMethod/AnInternalType;Ljava/lang/String;)V
  #35 = Utf8               returningInternal
  #36 = Utf8               (Lmock/plugin/internal/defaultMethod/NoInternalTypeUsageInterface;)Linternal/defaultMethod/AnInternalType;
  #37 = Utf8               (Linternal/defaultMethod/InterfaceWithDefaultMethodUsingInternalAPI;)Linternal/defaultMethod/AnInternalType;
  #38 = NameAndType        #35:#37        // returningInternal:(Linternal/defaultMethod/InterfaceWithDefaultMethodUsingInternalAPI;)Linternal/defaultMethod/AnInternalType;
  #39 = Methodref          #22.#38        // internal/defaultMethod/InterfaceWithDefaultMethodUsingInternalAPI$DefaultImpls.returningInternal:(Linternal/defaultMethod/InterfaceWithDefaultMethodUsingInternalAPI;)Linternal/defaultMethod/AnInternalType;
  #40 = Utf8               topInternal
  #41 = NameAndType        #40:#37        // topInternal:(Linternal/defaultMethod/InterfaceWithDefaultMethodUsingInternalAPI;)Linternal/defaultMethod/AnInternalType;
  #42 = Methodref          #22.#41        // internal/defaultMethod/InterfaceWithDefaultMethodUsingInternalAPI$DefaultImpls.topInternal:(Linternal/defaultMethod/InterfaceWithDefaultMethodUsingInternalAPI;)Linternal/defaultMethod/AnInternalType;
  #43 = Utf8               Lkotlin/Metadata;
  #44 = Utf8               mv
  #45 = Integer            1
  #46 = Integer            5
  #47 = Utf8               k
  #48 = Integer            3
  #49 = Utf8               xi
  #50 = Integer            48
  #51 = Utf8               DefaultImpls
  #52 = Utf8               mock/plugin/internal/defaultMethod/NoInternalTypeUsageInterface
  #53 = Class              #52            // mock/plugin/internal/defaultMethod/NoInternalTypeUsageInterface
  #54 = Utf8               NoInternalTypeUsageInterface.kt
  #55 = Utf8               Code
  #56 = Utf8               LineNumberTable
  #57 = Utf8               LocalVariableTable
  #58 = Utf8               RuntimeInvisibleAnnotations
  #59 = Utf8               RuntimeInvisibleParameterAnnotations
  #60 = Utf8               InnerClasses
  #61 = Utf8               SourceFile
  #62 = Utf8               RuntimeVisibleAnnotations
{
  public static internal.defaultMethod.AnInternalType internalArgsReturningInternal(mock.plugin.internal.defaultMethod.NoInternalTypeUsageInterface, internal.defaultMethod.AnInternalType, java.lang.String);
    descriptor: (Lmock/plugin/internal/defaultMethod/NoInternalTypeUsageInterface;Linternal/defaultMethod/AnInternalType;Ljava/lang/String;)Linternal/defaultMethod/AnInternalType;
    flags: (0x0009) ACC_PUBLIC, ACC_STATIC
    Code:
      stack=3, locals=3, args_size=3
         0: aload_1
         1: ldc           #10                 // String anInternalType
         3: invokestatic  #16                 // Method kotlin/jvm/internal/Intrinsics.checkNotNullParameter:(Ljava/lang/Object;Ljava/lang/String;)V
         6: aload_2
         7: ldc           #18                 // String s
         9: invokestatic  #16                 // Method kotlin/jvm/internal/Intrinsics.checkNotNullParameter:(Ljava/lang/Object;Ljava/lang/String;)V
        12: aload_0
        13: checkcast     #20                 // class internal/defaultMethod/InterfaceWithDefaultMethodUsingInternalAPI
        16: aload_1
        17: aload_2
        18: invokestatic  #25                 // Method internal/defaultMethod/InterfaceWithDefaultMethodUsingInternalAPI$DefaultImpls.internalArgsReturningInternal:(Linternal/defaultMethod/InterfaceWithDefaultMethodUsingInternalAPI;Linternal/defaultMethod/AnInternalType;Ljava/lang/String;)Linternal/defaultMethod/AnInternalType;
        21: areturn
      LineNumberTable:
        line 7: 12
        line 29: 21
      LocalVariableTable:
        Start  Length  Slot  Name   Signature
            0      22     0 $this   Lmock/plugin/internal/defaultMethod/NoInternalTypeUsageInterface;
            0      22     1 anInternalType   Linternal/defaultMethod/AnInternalType;
            0      22     2     s   Ljava/lang/String;
    RuntimeInvisibleAnnotations:
      0: #7()
        org.jetbrains.annotations.Nullable
    RuntimeInvisibleParameterAnnotations:
      parameter 0:
        0: #8()
          org.jetbrains.annotations.NotNull
      parameter 1:
        0: #8()
          org.jetbrains.annotations.NotNull
      parameter 2:
        0: #8()
          org.jetbrains.annotations.NotNull

  public static void internalArgsReturningVoid(mock.plugin.internal.defaultMethod.NoInternalTypeUsageInterface, internal.defaultMethod.AnInternalType, java.lang.String);
    descriptor: (Lmock/plugin/internal/defaultMethod/NoInternalTypeUsageInterface;Linternal/defaultMethod/AnInternalType;Ljava/lang/String;)V
    flags: (0x0009) ACC_PUBLIC, ACC_STATIC
    Code:
      stack=3, locals=3, args_size=3
         0: aload_1
         1: ldc           #10                 // String anInternalType
         3: invokestatic  #16                 // Method kotlin/jvm/internal/Intrinsics.checkNotNullParameter:(Ljava/lang/Object;Ljava/lang/String;)V
         6: aload_2
         7: ldc           #18                 // String s
         9: invokestatic  #16                 // Method kotlin/jvm/internal/Intrinsics.checkNotNullParameter:(Ljava/lang/Object;Ljava/lang/String;)V
        12: aload_0
        13: checkcast     #20                 // class internal/defaultMethod/InterfaceWithDefaultMethodUsingInternalAPI
        16: aload_1
        17: aload_2
        18: invokestatic  #34                 // Method internal/defaultMethod/InterfaceWithDefaultMethodUsingInternalAPI$DefaultImpls.internalArgsReturningVoid:(Linternal/defaultMethod/InterfaceWithDefaultMethodUsingInternalAPI;Linternal/defaultMethod/AnInternalType;Ljava/lang/String;)V
        21: return
      LineNumberTable:
        line 7: 12
        line 29: 21
      LocalVariableTable:
        Start  Length  Slot  Name   Signature
            0      22     0 $this   Lmock/plugin/internal/defaultMethod/NoInternalTypeUsageInterface;
            0      22     1 anInternalType   Linternal/defaultMethod/AnInternalType;
            0      22     2     s   Ljava/lang/String;
    RuntimeInvisibleParameterAnnotations:
      parameter 0:
        0: #8()
          org.jetbrains.annotations.NotNull
      parameter 1:
        0: #8()
          org.jetbrains.annotations.NotNull
      parameter 2:
        0: #8()
          org.jetbrains.annotations.NotNull

  public static internal.defaultMethod.AnInternalType returningInternal(mock.plugin.internal.defaultMethod.NoInternalTypeUsageInterface);
    descriptor: (Lmock/plugin/internal/defaultMethod/NoInternalTypeUsageInterface;)Linternal/defaultMethod/AnInternalType;
    flags: (0x0009) ACC_PUBLIC, ACC_STATIC
    Code:
      stack=1, locals=1, args_size=1
         0: aload_0
         1: checkcast     #20                 // class internal/defaultMethod/InterfaceWithDefaultMethodUsingInternalAPI
         4: invokestatic  #39                 // Method internal/defaultMethod/InterfaceWithDefaultMethodUsingInternalAPI$DefaultImpls.returningInternal:(Linternal/defaultMethod/InterfaceWithDefaultMethodUsingInternalAPI;)Linternal/defaultMethod/AnInternalType;
         7: areturn
      LineNumberTable:
        line 7: 0
        line 29: 7
      LocalVariableTable:
        Start  Length  Slot  Name   Signature
            0       8     0 $this   Lmock/plugin/internal/defaultMethod/NoInternalTypeUsageInterface;
    RuntimeInvisibleAnnotations:
      0: #7()
        org.jetbrains.annotations.Nullable
    RuntimeInvisibleParameterAnnotations:
      parameter 0:
        0: #8()
          org.jetbrains.annotations.NotNull

  public static internal.defaultMethod.AnInternalType topInternal(mock.plugin.internal.defaultMethod.NoInternalTypeUsageInterface);
    descriptor: (Lmock/plugin/internal/defaultMethod/NoInternalTypeUsageInterface;)Linternal/defaultMethod/AnInternalType;
    flags: (0x0009) ACC_PUBLIC, ACC_STATIC
    Code:
      stack=1, locals=1, args_size=1
         0: aload_0
         1: checkcast     #20                 // class internal/defaultMethod/InterfaceWithDefaultMethodUsingInternalAPI
         4: invokestatic  #42                 // Method internal/defaultMethod/InterfaceWithDefaultMethodUsingInternalAPI$DefaultImpls.topInternal:(Linternal/defaultMethod/InterfaceWithDefaultMethodUsingInternalAPI;)Linternal/defaultMethod/AnInternalType;
         7: areturn
      LineNumberTable:
        line 7: 0
        line 29: 7
      LocalVariableTable:
        Start  Length  Slot  Name   Signature
            0       8     0 $this   Lmock/plugin/internal/defaultMethod/NoInternalTypeUsageInterface;
    RuntimeInvisibleAnnotations:
      0: #7()
        org.jetbrains.annotations.Nullable
    RuntimeInvisibleParameterAnnotations:
      parameter 0:
        0: #8()
          org.jetbrains.annotations.NotNull
}
InnerClasses:
  public static final #51= #22 of #20;    // DefaultImpls=class internal/defaultMethod/InterfaceWithDefaultMethodUsingInternalAPI$DefaultImpls of class internal/defaultMethod/InterfaceWithDefaultMethodUsingInternalAPI
  public static final #51= #2 of #53;     // DefaultImpls=class mock/plugin/internal/defaultMethod/NoInternalTypeUsageInterface$DefaultImpls of class mock/plugin/internal/defaultMethod/NoInternalTypeUsageInterface
SourceFile: "NoInternalTypeUsageInterface.kt"
RuntimeVisibleAnnotations:
  0: #43(#44=[I#45,I#46,I#45],#47=I#48,#49=I#50)
    kotlin.Metadata(
      mv=[1,5,1]
      k=3
      xi=48
    )
```

</details>

Reply:
Done